VerilogCodingStyle.md

```diff
@@ -2291,6 +2292,109 @@ always_comb begin
 end
 ```
 
+### Functions and Tasks
+
+The following section applies to synthesizable RTL only.
+
+***In synthesizable RTL use of functions is allowed, provided they are declared
+automatic. Tasks should not be used.***
+
+Functions must be declared outside of a class in either a package or inside a
+module. A package is appropriate where the function relates to other definitions
+in the package and could be useful to multiple modules (even if its currently
+only used by one). A module is appropriate where the function specifically
+relates to the internals of that module.
+
+Functions should aim to conceptually represent a reusable block of combinational
+logic.
+
+Storage types must be explicitly declared for all arguments and the function
+return value.
+
+Do not use `input`, `output`, `inout`, or `ref` on function arguments.
+
+Do not assign to function arguments in the function body.
+
+&#x1f44d;
+```systemverilog {.good}
+function automatic logic [2:0] foo(logic [2:0] a, logic [2:0] b);
+ return a ^ b;
+endfunction
+```
+
+&#x1f44e;
+```systemverilog {.bad}
+function automatic [2:0] foo(input [2:0] a, [2:0] b);
+  b = b + 1;
+  return a + b;
+endfunction
+```
+
+It is recommended that data is returned from a function using an explicit
```

Comment:
(Copying comment as I updated the file and it doesn't appear on the new version)

I had initially said return style only but noticed that the aes block used function_name = result in some cases.

My main reason for disliking that style is it can somewhat obscure the control flow that leads you to the result, especially if you have code paths that give you two or more function_name = result statements.

Reply:
I see no good reason for the `function_name = result` style, so I'd completely disallow it and change the AES block.

Reply:
@vogelpi (or anyone else) do you wish to offer any defense for this style or are you happy for it to be disallowed?

Reply:
I wouldn't mind at all to change those functions in the AES module! I think having an explicit return value is cleaner and much more readable.

Reply:
Cool in that case I will switch it to allowing `return result` only.